### PR TITLE
Add optional raw camera pointer to image class

### PR DESCRIPTION
**Pull request by B1ueber2y**
Reverting to the design choice of https://github.com/colmap/colmap/pull/2697 while dropping the design from https://github.com/colmap/colmap/pull/2702.

With the camera pointer being optional, the best design IMO for the reconstruction would be still be keeping the ``SyncCameraPointers`` method. So the overall design from https://github.com/colmap/colmap/pull/2697 is kept here while only updating the shared pointer to raw pointer. 

Example following https://github.com/colmap/colmap/pull/2697:
```
import pycolmap
import numpy as np
recon = pycolmap.Reconstruction("./pycolmap/example/sfm/0")
image = recon.images[1]
print(image.camera) 
# Camera(camera_id=1, model=SIMPLE_RADIAL, width=3072, height=2048, params=[2776.773904, 1536.000000, 1024.000000, -0.003993] (f, cx, cy, k))
print(image.project_point(np.array([0., 0., 0.]))) 
# [-1684.30411175  1236.07593196]

print(image)
# Image(image_id=1, camera=Camera(camera_id=1), name="0000.png", triangulated=3620/9946)
print(image.has_camera(), image.has_camera_ptr())
# True True
image.reset_camera_ptr()
print(image)
# Image(image_id=1, camera_id=1, name="0000.png", triangulated=3620/9946)
print(image.has_camera(), image.has_camera_ptr())
# True False
```

### CODE REVIEW

File: src/colmap/scene/image.h [hunk 1]
```diff
@@ -149,11 +163,29 @@ inline camera_t Image::CameraId() const { return camera_id_; }
...
 
 inline bool Image::HasCamera() const { return camera_id_ != kInvalidCameraId; }
 
+inline struct Camera* Image::CameraPtr() const {
+  THROW_CHECK(HasCameraPtr());
+  return camera_;
+}
+
+inline void Image::SetCameraPtr(struct Camera* camera) {
+  THROW_CHECK_NE(camera->camera_id, kInvalidCameraId);
```
Comment (ahojnnes) [inline src/colmap/scene/image.h:180]: THROW_CHECK_NOTNULL(camera);

File: src/colmap/scene/image.h [hunk 2]
```diff
@@ -149,11 +163,29 @@ inline camera_t Image::CameraId() const { return camera_id_; }
...
+  camera_ = camera;
+  camera_id_ = camera->camera_id;
```
Comment (ahojnnes) [inline src/colmap/scene/image.h:182]: This is technically not needed anymore now?

File: src/colmap/scene/image.h [hunk 3]
```diff
@@ -149,11 +163,29 @@ inline camera_t Image::CameraId() const { return camera_id_; }
...
+}
+
+inline void Image::UnsetCameraPtr() { camera_ = nullptr; }
```
Comment (ahojnnes) [inline src/colmap/scene/image.h:185]: nit: ResetCameraPtr ?

File: src/colmap/scene/image.h (hunk 2 above)
Comment (ahojnnes) [inline src/colmap/scene/image.h:182]: THROW_CHECK_EQ(camera_id_, camera->camera_id) ?

Comment (ahojnnes) [reply src/colmap/scene/image.h:182]: Similarly, in the SetCameraId() method, we probably want to do `if (HasCameraPtr()) { THROW_CHECK_EQ(camera_id_, camera->camera_id)` ?

File: src/colmap/scene/reconstruction.h [hunk 4]
```diff
@@ -102,8 +102,8 @@ class Reconstruction {
   // might be taken by the same camera.
   void AddCamera(struct Camera camera);
 
-  // Add new image.
-  void AddImage(class Image image);
+  // Add new image. optionally load camera address in the image
```
Comment (ahojnnes) [inline src/colmap/scene/reconstruction.h:105]:
```suggestion
  // Add new image. Optionally load camera address in the image.
```
nit

File: src/pycolmap/scene/image.cc [hunk 5]
```diff
@@ -26,10 +26,15 @@ std::string PrintImage(const Image& image) {
   std::stringstream ss;
   ss << "Image(image_id="
      << (image.ImageId() != kInvalidImageId ? std::to_string(image.ImageId())
-                                            : "Invalid")
-     << ", camera_id="
-     << (image.HasCamera() ? std::to_string(image.CameraId()) : "Invalid")
-     << ", name=\"" << image.Name() << "\""
+                                            : "Invalid");
+  if (!image.HasCameraPtr()) {
```
Comment (ahojnnes) [inline src/pycolmap/scene/image.cc:30]: Assuming that the camera ptr and the id are always consistent, we would not need this branch here?

**Comment by B1ueber2y** [comment]
Once we are happy with the class design, I could try to start removing the use of ``colmap::CameraRig`` and visual-only ``RigBundleAdjuster`` with the new ``Frame`` class, as a better starting point for multi-sensor rig development.

File: src/colmap/scene/image.h (hunk 2 above)
Comment (B1ueber2y) [reply src/colmap/scene/image.h:183]: There might be some confusion here.

The idea of the design is that ``camera_`` always has higher priority than ``camera_id_``, so the ``camera_id`` should change accordingly to the corresponding camera. Thus, ``SetCameraId()`` is disabled when ``camera_`` is present (see here): https://github.com/colmap/colmap/pull/2715/files#diff-665386c3e6e1ebf0f2b3f383a2925c7a970b2e256f2f889f27fb090065158fbdR166

File: src/pycolmap/scene/image.cc (hunk 5 above)
Comment (B1ueber2y) [reply src/pycolmap/scene/image.cc:30]: Here I update the printing to show to the user if the optional camera ptr is set or not. Check the python example in the PR description (the printing is slightly different)

**Pull request closed without merging**